File: modules/events-mod/src/main/typescript/main/cache/EventDataCache.ts
import KafkatorioSettings from "../settings/KafkatorioSettings";
import {KafkatorioKeyedPacketData} from "../types";

export namespace EventUpdates {


  declare const global: {
    /** Map the hash of the key, as a string, a cache entry */
    cache: LuaTable<string, CacheEntry<KafkatorioKeyedPacketData>>,
  }


  export class Manager {


    init(force?: boolean) {

      const isAnythingUndefined = global.cache == undefined

      log(`Initialising EventDataCache (force=${force}, isAnythingUndefined=${isAnythingUndefined})...`)

      if (force == true || isAnythingUndefined) {
        log("Initialising global.cache")
        global.cache = new LuaTable<string, CacheEntry<KafkatorioKeyedPacketData>>()
      }
      log(`Finished initialising EventDataCache`)
    }


    /**
     * Update the cache data without restarting the 'time-to-emit' countdown, so that the data won't
     * be emitted more than once per {@link CacheEntry.expirationDurationTicks}.
     */
    public throttle<PACKET extends KafkatorioKeyedPacketData>(
        key: PacketKey<PACKET>,
        type: PacketType<PACKET>,
        mutate: CacheDataMutator<PACKET>,
        expirationDurationTicks?: uint,
    ) {
      this.update(key, type, mutate, false, expirationDurationTicks)
    }


    /**
     * Update the cache data and reset the 'time-to-emit' countdown, so the data won't be emitted
     * until there's an inactivity gap of {@link CacheEntry.expirationDurationTicks}.
     */
    public debounce<PACKET extends KafkatorioKeyedPacketData>(
        key: PacketKey<PACKET>,
        type: PacketType<PACKET>,
        mutate: CacheDataMutator<PACKET>,
        expirationDurationTicks?: uint,
    ) {
      this.update(key, type, mutate, true, expirationDurationTicks)
    }


    private update<PACKET extends KafkatorioKeyedPacketData>(
        key: PacketKey<PACKET>,
        type: PacketType<PACKET>,
        mutate: CacheDataMutator<PACKET>,
        resetLastUpdated: boolean,
        expirationDurationTicks?: uint,
    ) {
      const entry: CacheEntry<PACKET> = this.getCacheEntry(key, type) ??
                                        Manager.createCacheEntry(key, type)
      mutate(entry.packet)
      if (resetLastUpdated) {
        entry.lastUpdatedTick = game.tick
      }
      if (expirationDurationTicks != undefined) {
        entry.expirationDurationTicks = expirationDurationTicks
      }
    }


    /**
     * Usage example: If an urgent update comes in, then the expiration can be reduced so that
     * it's transmitted sooner.
     */
    public setExpiration<PACKET extends KafkatorioKeyedPacketData>(
        key: PacketKey<PACKET>,
        type: PacketType<PACKET>,
        expirationDurationTicks: uint,
    ) {
      const entry = this.getCacheEntry(key, type)
      if (entry != undefined) {
        entry.expirationDurationTicks = expirationDurationTicks
      }
    }


    // use in/out v4.7.0 https://github.com/microsoft/TypeScript/pull/48240
    public extractExpired(): Array<KafkatorioKeyedPacketData> {
      let countExpired = 0
      let countTotal = 0
      const expiredData: Array<KafkatorioKeyedPacketData> = []
      for (let [key, entry] of global.cache) {
        countTotal++
        if (this.isExpired(entry)) {
          countExpired++
          expiredData.push(entry.packet)
          global.cache.delete(key)
        }
      }

      const hasExpiredItems = countExpired > 0
      const hasAnyItem = countTotal > 0

      if (hasExpiredItems || (game.tick % 30 == 0 && hasAnyItem)) {
        log(`[extractExpired] expired items: ${countExpired}, total: ${countTotal}`)
      }
      return expiredData
    }


    private getCacheEntry<PACKET extends KafkatorioKeyedPacketData>(
        key: EventUpdates.PacketKey<PACKET>,
        type: EventUpdates.PacketType<PACKET>,
    ): CacheEntry<PACKET> | undefined {
      const hash = Manager.hashKey(key)
      if (!global.cache.has(hash)) {
        return undefined
      }
      const value: CacheEntry<any> = global.cache.get(hash)
      if (this.isEntryInstanceOf(value, type)) {
        return value
      } else {
        // Type mismatch. This shouldn't happen...
        return undefined
      }
    }


    private static createCacheEntry<PACKET extends KafkatorioKeyedPacketData>(
        key: PacketKey<PACKET>,
        type: PacketType<PACKET>,
    ): CacheEntry<PACKET> {
      const data: PACKET = <PACKET>{
        type: type,
        key: key,
      }
      const entry: CacheEntry<PACKET> = new CacheEntry<PACKET>(data)
      const hash = Manager.hashKey(key)
      global.cache.set(hash, entry)
      return entry
    }


    isExpired<TYPE extends KafkatorioKeyedPacketData>(entry: CacheEntry<TYPE>): boolean {
      const expiryDuration =
          entry.expirationDurationTicks
          ?? KafkatorioSettings.getEventCacheExpirationTicks(entry.packet)

      if (expiryDuration == undefined) {
        return true
      } else {
        return game.tick - entry.lastUpdatedTick > expiryDuration
      }
    }


    isEntryInstanceOf<PACKET extends KafkatorioKeyedPacketData>(
        entry: CacheEntry<any> | undefined,
        type: EventUpdates.PacketType<PACKET>,
    ): entry is CacheEntry<PACKET> {
      return this.isDataInstanceOf(entry?.packet, type)
    }


    isDataInstanceOf<PACKET extends KafkatorioKeyedPacketData>(
        packet: PACKET | undefined,
        type: EventUpdates.PacketType<PACKET>,
    ): packet is PACKET {
      return packet != undefined && packet.type == type
    }

    private static hashKey<PACKET extends KafkatorioKeyedPacketData>(
        key: PacketKey<PACKET>,
    ): string {
      return game.encode_string(game.table_to_json(key))!!
    }


    // function isKeyInstanceOf<TYPE extends FactorioEventUpdateType>(
    //     key: CacheKey<any> | undefined,
    //     updateType: TYPE,
    // ): key is CacheKey<TYPE> {
    //   return key != undefined && key.updateType == updateType
    // }

  }


  // export type KeyedPacketType = Exclude<KafkatorioPacketDataType,
  //     "CONFIG" | "CONSOLE_CHAT" | "CONSOLE_COMMAND" | "PROTOTYPES" | "SURFACE">

  // /** Map a {@link KeyedPacketType} to a {@link KafkatorioPacketData} DTO */
  // type ConvertToUpdate<TYPE extends KafkatorioKeyedPacketData> =
  //     TYPE extends "PLAYER" ? PlayerUpdate :
  //     TYPE extends "MAP_CHUNK" ? MapChunkUpdate :
  //     TYPE extends "ENTITY" ? EntityUpdate :
  //     never


  // /** Map a {@link KeyedPacketType} to a {@link KafkatorioKeyedPacketKey} */
  // type ConvertToUpdateKey<TYPE extends KafkatorioKeyedPacketData> =
  //     TYPE extends "PLAYER" ? PlayerUpdateKey :
  //     TYPE extends "MAP_CHUNK" ? MapChunkUpdateKey :
  //     TYPE extends "ENTITY" ? EntityUpdateKey :
  //     never


  // /** Make `updateType` specific */
  export type PacketKey<PACKET extends KafkatorioKeyedPacketData> = PACKET["key"]
  export type PacketType<PACKET extends KafkatorioKeyedPacketData> = PACKET["type"]
  //
  //
  // /** Make `updateType` specific */
  // export type CacheData<TYPE extends KafkatorioKeyedPacketData> =
  //     CacheTyped<TYPE>
  //     & Omit<ConvertToUpdate<TYPE>, "updateType">
  //
  //
  // export type CacheTyped<TYPE extends KafkatorioKeyedPacketData> = {
  //   readonly updateType: TYPE
  // }


  export type CacheDataMutator<PACKET extends KafkatorioKeyedPacketData> = (packet: PACKET) => void


  class CacheEntry<PACKET extends KafkatorioKeyedPacketData> {
    lastUpdatedTick: uint = game.tick
    expirationDurationTicks?: uint

    packet: PACKET

    constructor(
        packet: PACKET,
        expirationDurationTicks?: uint,
    ) {

      this.packet = packet
      this.expirationDurationTicks = expirationDurationTicks
    }
  }


}

const EventUpdatesManager = new EventUpdates.Manager()

export default EventUpdatesManager

File: modules/events-mod/src/main/lua/main/cache/EventDataCache.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____KafkatorioSettings = require("main.settings.KafkatorioSettings")
local KafkatorioSettings = ____KafkatorioSettings.default
local ____PacketEmitter = require("main.PacketEmitter")
local packetEmitter = ____PacketEmitter.default
____exports.EventUpdates = {}
local EventUpdates = ____exports.EventUpdates
do
    local CacheEntry
    EventUpdates.Manager = __TS__Class()
    local Manager = EventUpdates.Manager
    Manager.name = "Manager"
    function Manager.prototype.____constructor(self)
    end
    function Manager.prototype.init(self, force)
        local isAnythingUndefined = global.cache == nil
        log(((("Initialising EventDataCache (force=" .. tostring(force)) .. ", isAnythingUndefined=") .. tostring(isAnythingUndefined)) .. ")...")
        if force == true or isAnythingUndefined then
            log("Initialising global.cache")
            global.cache = {}
        end
        log("Finished initialising EventDataCache")
    end
    function Manager.prototype.throttle(self, key, ____type, mutate, expirationDurationTicks)
        self:update(
            key,
            ____type,
            mutate,
            false,
            expirationDurationTicks
        )
    end
    function Manager.prototype.debounce(self, key, ____type, mutate, expirationDurationTicks)
        self:update(
            key,
            ____type,
            mutate,
            true,
            expirationDurationTicks
        )
    end
    function Manager.prototype.update(self, key, ____type, mutate, resetLastUpdated, expirationDurationTicks)
        local entry = self:getCacheEntry(key, ____type) or EventUpdates.Manager:createCacheEntry(key, ____type)
        mutate(entry.packet)
        if resetLastUpdated then
            entry.lastUpdatedTick = game.tick
        end
        if expirationDurationTicks ~= nil then
            entry.expirationDurationTicks = expirationDurationTicks
        end
        if entry.expirationDurationTicks ~= nil and entry.expirationDurationTicks <= 0 then
            self:extractAndEmitExpiredPackets()
        end
    end
    function Manager.prototype.setExpiration(self, key, ____type, expirationDurationTicks)
        local entry = self:getCacheEntry(key, ____type)
        if entry ~= nil then
            entry.expirationDurationTicks = expirationDurationTicks
        end
    end
    function Manager.prototype.extractExpired(self)
        local countExpired = 0
        local countTotal = 0
        local expiredData = {}
        for key, entry in pairs(global.cache) do
            countTotal = countTotal + 1
            if self:isExpired(entry) then
                countExpired = countExpired + 1
                expiredData[#expiredData + 1] = entry.packet
                global.cache[key] = nil
            end
        end
        local hasExpiredItems = countExpired > 0
        local hasAnyItem = countTotal > 0
        if hasExpiredItems or game.tick % 30 == 0 and hasAnyItem then
            log((("[extractExpired] expired items: " .. tostring(countExpired)) .. ", total: ") .. tostring(countTotal))
        end
        return expiredData
    end
    function Manager.prototype.extractAndEmitExpiredPackets(self)
        local cachedEvents = self:extractExpired()
        if #cachedEvents > 0 then
            log((("[EventDataCache] emitting " .. tostring(#cachedEvents)) .. " events on tick ") .. tostring(game.tick))
            for ____, event in ipairs(cachedEvents) do
                packetEmitter:emitKeyedPacket(event)
            end
        end
        return #cachedEvents
    end
    function Manager.prototype.getCacheEntry(self, key, ____type)
        local hash = EventUpdates.Manager:hashKey(key)
        if not (global.cache[hash] ~= nil) then
            return nil
        end
        local value = global.cache[hash]
        if self:isEntryInstanceOf(value, ____type) then
            return value
        else
            return nil
        end
    end
    function Manager.createCacheEntry(self, key, ____type)
        local data = {type = ____type, key = key}
        local entry = __TS__New(CacheEntry, data)
        local hash = EventUpdates.Manager:hashKey(key)
        global.cache[hash] = entry
        return entry
    end
    function Manager.prototype.isExpired(self, entry)
        local expiryDuration = entry.expirationDurationTicks or KafkatorioSettings:getEventCacheExpirationTicks(entry.packet)
        if expiryDuration == nil then
            return true
        else
            return game.tick - entry.lastUpdatedTick > expiryDuration
        end
    end
    function Manager.prototype.isEntryInstanceOf(self, entry, ____type)
        local ____self_2 = self
        local ____self_isDataInstanceOf_3 = self.isDataInstanceOf
        local ____entry_packet_0 = entry
        if ____entry_packet_0 ~= nil then
            ____entry_packet_0 = ____entry_packet_0.packet
        end
        return ____self_isDataInstanceOf_3(____self_2, ____entry_packet_0, ____type)
    end
    function Manager.prototype.isDataInstanceOf(self, packet, ____type)
        return packet ~= nil and packet.type == ____type
    end
    function Manager.hashKey(self, key)
        return game.encode_string(game.table_to_json(key))
    end
    CacheEntry = __TS__Class()
    CacheEntry.name = "CacheEntry"
    function CacheEntry.prototype.____constructor(self, packet, expirationDurationTicks)
        self.lastUpdatedTick = game.tick
        self.packet = packet
        self.expirationDurationTicks = expirationDurationTicks
    end
end
local EventUpdatesManager = __TS__New(____exports.EventUpdates.Manager)
____exports.default = EventUpdatesManager
return ____exports
